doc/designs/fips_deferred_tests.md

```diff
@@ -0,0 +1,291 @@
+   for the lower-level algorithm should be triggered automatically if it has
+   not already been run.
+6. **Thread Safety**: The mechanism for checking test status and running
+   tests must be thread-safe.
+7. **Equivalency**: If running a KAT for one algorithm also satisfies the
+   testing requirements for other algorithms, a mechanism must exist to mark
+   all equivalent tests as passed.
+
+Self-Test State Management
+--------------------------
+
+A basic mechanism to mark individual tests for deferral and track their state
+has been introduced with PR #28725 and will be used as a basis for
+implementing a more complete deferral and dependency tracking mechanism as
+described in the requirements.
+
+In particular, the state management will be expanded to include equivalency
+and explicit dependency requirements.
+
+The FIPS_kat_deferred() function will handle recursing into executing explicit
+dependencies if needed, and marking all equivalent tests as passed at once.
+
+This function operates on the basis of locking execution of tests to a single
+thread and therefore can safely manipulate the data required to maintain
+state across all threads as well as handling ordering of test execution.
+
+Triggering Self-Tests
+---------------------
+
+Each algorithm implementation will be modified to trigger its corresponding
+self-test at initialization similar to how SLH-DSA tests have already been
+changed. This will typically be within the algorithm's `newctx()` or `init()`
+function, as this is the earliest point at which an application signals its
+intent to use the algorithm.
+
+The self-test state is maintained in a structure of type `FIPS_DEFERRED_TEST`,
+this structure will be augmented with two arrays of pointers to other state
+structures of the same type.
+
+One will handle test equivalency and will be named 'also_satisfies', the other
+will handle explicit dependencies that are not implicitly handled by
+initialization functions.
+
+Note that the `also_satisfies` list will *not* be used recursively (see the
+Examples section for an explanation of why this can't be done), while the
+`depends_on` list is used in a recursive fashion.
+
+Dependency Handling
+-------------------
+
+"Implicit" dependencies will be managed by the self-test functions themselves.
+By simply calling the initialization method of functions that are depended on,
+the self-test of the dependent algorithm will be triggered. This will lead to
+a call to the `FIPS_kat_deferred()` function which will recursively execute
+only tests that are not already marked in progress. Therefore this call can
+safely nest calls to self-tests without risking infinite loops.
+
+OpenSSL functional tests may need to be changed to account for test nesting,
+especially when fault injection is implemented.
+
+In case a dependency is not implicit, an explicit link to a different test
+state will be added to the `depends_on` list.
```

Comment:
Is there a case where `depends_on` would be more than one other test?
I.e. does it need to be a list?